build: Fix coverage builds

hebasto opened the pull request:
This PR follows up on https://github.com/bitcoin/bitcoin/pull/30811, which inadvertently broke coverage builds:
1. For GCC. See https://github.com/bitcoin/bitcoin/pull/31337#issuecomment-2490598011.
2. For [Clang's source-based code coverage](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html) in the OSS-Fuzz environment due to its use of other options and a third party script. See https://issues.oss-fuzz.com/issues/379122777.

The root cause of this regression is that the `-ffile-prefix-map` option implicitly applies:
-  [`-fprofile-prefix-map`](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#index-fprofile-prefix-map) when using GCC.
- [`-fcoverage-prefix-map`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fcoverage-prefix-map) when using Clang. ~This behaviour is not explicitly documented~ See https://github.com/llvm/llvm-project/commit/994c544c18c86cbdb6536aae5d27ef7e2f592486.

With this PR, only the `-fdebug-prefix-map` and `-fmacro-prefix-map` options are applied.

**Note for reviewers:** Please ensure that https://github.com/bitcoin/bitcoin/issues/30799 is not reintroduced.

maflcko reviewed (commented):
Good catch, but I think the upstream docs were wrong in this case. For me, they don't mention coverage is affected as well: https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-ffile-prefix-map

fanquake commented:
As far as I'm aware, Clang based coverage builds are working outside of oss-fuzz (cc @dergoegge), so it's not clear why this would be the right solution. If it's an oss-fuzz only problem, shouldn't the fix happen there?

> With this PR, only the -fdebug-prefix-map and -fmacro-prefix-map options are applied only for non-release builds.

Not sure about introducing these kinds of (undocumented) distinctions. By release do you just mean Guix build? Why should these options only apply to one build and not the other (regardless of anything oss-fuzz related).

fanquake commented on CMakeLists.txt:
```diff
@@ -436,8 +436,11 @@ configure_file(contrib/filter-lcov.py filter-lcov.py USE_SOURCE_PERMISSIONS COPY
 
-try_append_cxx_flags("-ffile-prefix-map=A=B" TARGET core_interface SKIP_LINK
-  IF_CHECK_PASSED "-ffile-prefix-map=${PROJECT_SOURCE_DIR}/src=."
+try_append_cxx_flags("-fdebug-prefix-map=A=B" TARGET core_interface SKIP_LINK
```
If this was going to be done, this would need some sort of documentation, otherwise someone would likely assume this can be consolidated back to `-ffile-prefix-map`.

hebasto replied:
The comment has been added.

dergoegge commented:
Here is a recent example of clang coverage working for me: http://bitcoind-fuzz.dergoegge.de:8000/bitcoin/harnesses/rpc/coverage_report/coverage/workdir/bitcoin/index.html

I'm slightly confused why oss-fuzz broke while my setup keeps working. Afaict I'm using the same flags and `llvm-*` tools as them...

hebasto commented:
> I'm slightly confused why oss-fuzz broke while my setup keeps working. Afaict I'm using the same flags and `llvm-*` tools as them...

They [use](https://github.com/google/oss-fuzz/blob/51922e43fa80ae858fcd12aa84be3c0c75ed7b0d/infra/base-images/base-runner/coverage#L54-L55) the `-path-equivalence=...` option:
```bash
# Use path mapping, as $SRC directory from the builder is copied into $OUT/$SRC.
PATH_EQUIVALENCE_ARGS="-path-equivalence=/,$OUT"
```

hebasto committed:
build: Avoid using the `-ffile-prefix-map` compiler option

The `-ffile-prefix-map` compiler option implies `-fprofile-prefix-map`
on GCC or `-fcoverage-prefix-map` on Clang, which can lead to issues
with coverage builds.

This change applies only the options necessary for build reproducibility
and accurate source location messages.

hebasto commented:
Reworked per feedback.

fanquake commented:
> which inadvertently broke coverage builds for both [Clang's source-based code coverage](https://issues.oss-fuzz.com/issues/379122777)

Can you clarify that this is only for oss-fuzz, due to its use of other options and a third party script, otherwise, source based code coverage using Clang currently still works fine with master (see: https://github.com/bitcoin/bitcoin/pull/31337#issuecomment-2490620693). Similarly, in the inline comment you've added, can you be a bit more specific than "can cause issues"; as we seemingly know what the issues are.

hebasto commented on contrib/guix/libexec/build.sh:
```diff
@@ -212,6 +212,7 @@ CONFIGFLAGS="-DREDUCE_EXPORTS=ON -DBUILD_BENCH=OFF -DBUILD_GUI_TESTS=OFF -DBUILD
 HOST_CFLAGS="-O2 -g"
 HOST_CFLAGS+=$(find /gnu/store -maxdepth 1 -mindepth 1 -type d -exec echo -n " -ffile-prefix-map={}=/usr" \;)
 case "$HOST" in
+    *linux*)  HOST_CFLAGS+=" -ffile-prefix-map=${DISTSRC}/src=." ;;
```
[Reworked](https://github.com/bitcoin/bitcoin/pull/31337#issuecomment-2490932956).

hebasto commented:
> > which inadvertently broke coverage builds for both [Clang's source-based code coverage](https://issues.oss-fuzz.com/issues/379122777)
> 
> Can you clarify that this is only for oss-fuzz, due to its use of other options and a third party script, otherwise, source based code coverage using Clang currently still works fine with master (see: [#31337 (comment)](https://github.com/bitcoin/bitcoin/pull/31337#issuecomment-2490620693)). Similarly, in the inline comment you've added, can you be a bit more specific than "can cause issues"; as we seemingly know what the issues are.

Thanks! Updated.

hebasto commented:
> > source based code coverage using Clang currently still works fine with master (see: [#31337 (comment)](https://github.com/bitcoin/bitcoin/pull/31337#issuecomment-2490620693)).
> 
> Would be nice to do it for this pull as well, to see if there is any difference.

cc @dergoegge

dergoegge reviewed (approved):
tACK 01a7298818d7bd66bb8c800d7cd30ae0a4d3b1d1

dergoegge commented:
> Is this fix related to the "no coverage data" results in https://corecheck.dev/ (i.e. https://corecheck.dev/bitcoin/bitcoin/pulls/31337), or is that a different issue?

I think the reason for "no coverage data" on corecheck's site for this PR is that this PR doesn't change any source files covered by any tests. It does show coverage changes under "Lost/Gained baseline coverage" but those are likely caused by non-determinism in the tests.

This PR was triggered by coverage builds on oss-fuzz being broken.

fanquake commented on CMakeLists.txt:
```diff
@@ -436,8 +436,16 @@ configure_file(contrib/filter-lcov.py filter-lcov.py USE_SOURCE_PERMISSIONS COPY
+# Avoiding the `-ffile-prefix-map` compiler option because it implies
+# `-fcoverage-prefix-map` on Clang or `-fprofile-prefix-map` on GCC,
+# which can cause issues with coverage builds, particularly when using
+# Clang in the OSS-Fuzz environment due to its use of other options
```
I'm not sure this is a good inline comment, because it doesn't actually document the problem, just the side-effects.

hebasto replied:
Mind suggesting a better wording for the comment?

fanquake replied:
I don't really have one, given it's a somewhat obscure issue, but which is also why it'd be good to have proper documentation for. Anyone reading this in future isn't going to have any information other than, "don't use ffile-prefix-map otherwise it might break something in oss-fuzz". However given how long the build has been broken now, going to merge this to unbreak it.
